mgmt: mcumgr: grp: img_mgmt: Add optional max image size reduction

nordicjm opened the pull request:
Adds an optional feature that can be used to reduce the maximum allowed image upload file size whereby an image could be uploaded that would be too large to swap even if it could fit the partition

Fixes #46194

nordicjm committed:
mgmt: mcumgr: grp: img_mgmt: Add optional max image size reduction

Adds an optional feature that can be used to reduce the maximum
allowed image upload file size whereby an image could be uploaded
that would be too large to swap even if it could fit the partition

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:
```diff
@@ -153,6 +153,38 @@ config MCUMGR_GRP_IMG_MUTEX
+	  configuration, otherwise an image can be uploaded that is too large for the bootloader
+	  to swap, this selects which method to use.
+
+	  Note: setting this to a non-disabled option will prevent uploading of padded and
+	  confirmed images, if support for that is required then this feature should be left as
+	  disabled.
```
Would it be possible for us to develop MCUboot/bootutil function that would check the image for valid footer?

nordicjm replied:
Not possible, well, depending on what you mean by valid footer. Only MCUboot can confirm the image properly due to possibly being encrypted, an application could do some minor validation but nothing more than that

de-nordic replied:
Actually you can use imgtool to pad, confirm and encrypt image. 
But yes, I guess you can not check the footer if the MCUboot only knows the key.

nordicjm replied:
Also can't check that here because to check that you need to receive the whole image

de-nordic replied:
The image slot trailer, the one that stores swap bits, etc, is not encrypted. actually.

nordicjm replied:
Not encrypted no but only mcuboot can verify that it is a valid image or supported on that device, the application can't know anything more than the basic validity of parts of the image

de-nordic commented on subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:
```diff
@@ -153,6 +153,38 @@ config MCUMGR_GRP_IMG_MUTEX
+	help
+	  Will use the image overhead size calculated during Sysbuild image configuration.
+
+config MCUMGR_GRP_IMG_TOO_LARGE_BOOTLOADER_INFO
+	bool "Via retention bootloader info"
+	depends on RETENTION_BOOTLOADER_INFO_OUTPUT_FUNCTION
+	help
+	  Will fetch the maximum image size from the bootloader info retention subsystem module.
```
We should be able to, also, provide this info at compile time. At all we are able to calculate space needed for footer in MCUboot, when we know size of slot and write block size.

nordicjm replied:
This is provided at compile time when using the sysbuild option. If not using sysbuild then you cannot get that information

nordicjm committed:
mgmt: mcumgr: grp: os_mgmt: Add firmware uploader boot type

Adds firmware uploader to the output of bootloader mode for
MCUboot

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
dfu: Add support for MCUboot estimated update image overhead size

Adds support for an overhead size which MCUboot can set when using
sysbuild, this can be used to check the provided size of an
application being uploaded to ensure it will fit and swap without
being rejected

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  a4eda30f5b0cfd0cf15512be9dcd559239dbfc91

Brings following Zephyr relevant fixes:
 - a4eda30f zephyr: Add estimated size of update trailer to
   sysbuild
 - 205d7e5b boot_serial: Adapt to zcbor 0.8.x

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
